[analyzer] Cppcheck support (#3680)
Ericsson/codechecker: docs/analyzer/checker_and_analyzer_configuration.md

```
-config="{ "Checks": "clang-diagnostic-*,clang-analyzer-*", "WarningsAsErrors": "", "HeaderFilterRegex": "", "AnalyzeTemporaryDtors": false, "CheckOptions": [ { "key": "google-readability-braces-around-statements.ShortStatementLines", "value": "1" }, { "key": "modernize-loop-convert.MaxCopySize", "value": "16" }, { "key": "modernize-loop-convert.NamingStyle", "value": "CamelCase" }, { "key": "modernize-use-nullptr.NullMacros", "value": "NULL" } ] }"
```

# Cppcheck

As of CodeChecker 6.20, Codechecker can now execute the Cppcheck analyzer.

## Configuration

> The Cppcheck analyzer can be configured with --analyzer-config cppcheck:* paramterers.

Reviewer: ```suggestion
The Cppcheck analyzer can be configured with `--analyzer-config cppcheck:*` paramterers.
```

> The supporte analyzer configuration items can be listed with `CodeChecker analyzers --analyzer-config cppcheck --details`

Reviewer: ```suggestion
The supported analyzer configuration items can be listed with `CodeChecker analyzers --analyzer-config cppcheck --details`
```

> As of the CodeChecker 6.20, the following options are supported:

Reviewer: ```suggestion
As of CodeChecker 6.20, the following options are supported:
```

* `cppcheck:addons` A list of cppcheck addon files.
* `cppcheck:libraries` A list of cppcheck library definiton files.
* `cppcheck:platform` The platform configuration .xml file.

> Please note that for addons and libraries, multiple items can be specified in the following format: --analyzer-config cppcheck:addons <addon.py> --analyzer-config cppcheck:addons <addon2.py>.

Reviewer: ```suggestion
Please note that for addons and libraries, multiple items can be specified in the following format: `--analyzer-config cppcheck:addons <addon.py> --analyzer-config cppcheck:addons <addon2.py>`.
```

Cppcheck only supports a limited number of platforms. Custom bit lengths can be specified with a platform file.

> An example platform file from the cppcheck manual:

Reviewer: ```suggestion
An example platform file from the Cppcheck manual:
```
Lower and upper case versions of Cppcheck are mixed throughout in the documentation. This should be common for every occurrence.

Author: capitalized them

``` xml
<?xml version="1"?>
<platform>
  <char_bit>8</char_bit>
  <default-sign>signed</default-sign>
  <sizeof>
    <short>2</short>
    <int>4</int>
    <long>4</long>
    <long-long>8</long-long>
    <float>4</float>
    <double>8</double>
    <long-double>12</long-double>
    <pointer>4</pointer>
    <size_t>4</size_t>
    <wchar_t>2</wchar_t>
  </sizeof>
</platform>

> ```

Reviewer 2: Are the "common" platforms like LP64 (Unix) and LLP64 (Windows) modelled by Cppcheck by default? Or should we always generate this file if the user does not specify one? In the latter case, the generator code for this feature seems to be missing.

Author: The native platform is assumed if the config file is not provided.

## Limitations

The following limitations need to be considered when using Cppcheck:

* The whole program analysis feature of cppcheck is not supported. Cppcheck is invoked for every item in the provided compilation database.

>   * The unususedFunction checker of cppcheck is always disabled by default.

Reviewer: ```suggestion
  * The `cppcheck-unususedFunction` checker of Cppcheck is always disabled by default.
```

  * The CTU functionality of cppcheck is not supported.

> * The severity categorizations are only provided for the built in checkers. Addon checkeckers can be used, but their report's severity will be displayed as`Unspecified`.

Reviewer: ```suggestion
* The severity categorizations are only provided for the built-in checkers. Addon checkers can be used, but their reports' severity will be displayed as `Unspecified`.
```

> * The severity categorizations are only provided for the built in checkers. Addon checkeckers can be used, but their report's severity will be displayed as`Unspecified`.

Reviewer 2: If there are specific documentation pages for these checks and "CTU functionality", we should link it for further reading.

Author: I could not find any checker documentation on the internet, There is a manual for Cppcheck, but i don't think the limitations list is a good place to link the documentation.

> * The Cppcheck categorization of checkers is not yet introduced into the Cppcheck label file. To enable a whole category, each individual checker needs to be enabled with the `--enable` flag.

Reviewer: ```suggestion
* The Cppcheck categorization of checkers is not yet introduced into the [Cppcheck label file](config/labels/analyzers/cppcheck.json). To enable a whole category, each individual checker needs to be enabled with the `--enable` flag.
```
I'm not sure this link is correct with this relative path, please, make sure that it points to the .json file.

Reviewer 2: These links will be broken when the documentation is generated by mkdocs/readthedocs, however. It'll only look good when viewed in GitHub's viewer! #3486

Author: What should we do? Are the readthedocs maintained?

Reviewer: I wrote "I'm not sure" about the correctness of this path because I haven't tested it when I wrote this suggestion. I don't know what is the basis of such a relative path. I tried to find another example where we're referring another .md file, and I think this should rather be absolute, otherwise it's considered to be relative to the current .md file.

So I think we should use this absolute path:
/config/labels/analyzers/cppcheck.json

Reviewer 2: > What should we do? Are the readthedocs maintained?

@vodorok I am not sure if anything is maintained :innocent:, but with `@bruntib`'s suggestion, even if someone clicks the link and ends up in a broken place, the broken path will be shorter and easier to reason about!

* All Cppcheck Errors and Warnings are enabled by default.

> * If not configured, `Native` platform will be assumed for the analyzed compilation database. No platform translation will occur by CodeChecker. If other is needed, please provide a platform file with the correct bit lengths.

Reviewer: ```suggestion
* If not configured, `Native` platform will be assumed for the analyzed compilation database (i.e. the type sizes of the host system are used). No platform translation will occur by CodeChecker. If another one is needed, please provide a platform file with the correct bit lengths.
```

> * To reach maximum compatibility with the existing CodeChecker invocation, Cppcheck is ran with the `--enable=all` parameter, and all non needed checkers are passed in as `--suppress=<checker>`.

Reviewer: ```suggestion
* To reach maximum compatibility with the existing CodeChecker invocation, Cppcheck is run with the `--enable=all` parameter, and all non-needed checkers are passed in as `--suppress=<checker>`.
```

> * Due to legal reasons, no Misra rule texts are supplied.

Reviewer 2: ```suggestion
* Due to legal reasons, no MISRA rule texts are supplied.
```

What does this mean? We do not run these checks but Cppcheck would support them?

## Example invocation

> ``` shell

Reviewer 2: (Is this okay, with the space there?)

> CodeChecker check -l ./compile_commands.json \
> --analyzers cppcheck \ # Run Cppcheck analyzer only
> -e cppcheck-missingOverride \ # enable the missingOverride cppcheck check
> -d cppcheck-virtualCallInConstructor \ # disable the virtualCallInConstructor check
> --analyzer-config cppcheck:addons=../cppcheck/addons/misc.py \ # enable the misc checks
> --analyzer-config cppcheck:addons=../cppcheck/addons/cert.py \ # enable cert cheks
> --analyzer-config cppcheck:libraries=../cppcheck/cfg/zlib.cfg \ # add zlib definitons
> --analyzer-config cppcheck:libraries=../cppcheck/cfg/gnu.cfg \ # add gnu definitions
> -o ./reports

Reviewer 2: ```suggestion
CodeChecker check -l ./compile_commands.json \
  --analyzers cppcheck \ # Run Cppcheck analyzer only
  -e cppcheck-missingOverride \ # enable the missingOverride cppcheck check
  -d cppcheck-virtualCallInConstructor \ # disable the virtualCallInConstructor check
  --analyzer-config cppcheck:addons=../cppcheck/addons/misc.py \ # enable the misc checks
  --analyzer-config cppcheck:addons=../cppcheck/addons/cert.py \ # enable cert cheks
  --analyzer-config cppcheck:libraries=../cppcheck/cfg/zlib.cfg \ # add zlib definitions
  --analyzer-config cppcheck:libraries=../cppcheck/cfg/gnu.cfg \ # add gnu definitions
  -o ./reports
```
